### PR TITLE
Allows trailing comma and skipping arguments in array destructing and enables heredoc as function argument

### PR DESCRIPTION
**Pull request by cfroystad**
After adding a few more test projects, a few more edge cases of the grammar was identified. This PR adds those test projects, fixes any surfaced errors and adds tree-sitter tests for them explicitly.

### Changes

- Added test projects: Laravel Framework, MediaWiki
- Array destructuring changes:
  - Added support for trailing comma
  - Added support for skipping arguments
  - Require at least one argument
- Added support for using Heredoc as function argument

Checklist:
- [x] All tests pass in CI
- [x] There are sufficient tests for the new fix/feature (new tests added for edge cases)
- [x] Grammar rules have not been renamed unless absolutely necessary (0 rules renamed)
- [x] The conflicts section hasn't grown too much (0 new conflicts)
- [x] The parser size hasn't grown too much (master: 2552, PR: 2579)

### CODE REVIEW

File: grammar.js [hunk 1]
```diff
@@ -1010,19 +1014,21 @@ module.exports = grammar({
     _list_destructing: $ => seq(
       'list',
       '(',
-      commaSep(optional(choice(
+      commaSep1(optional(choice(
         choice(alias($._list_destructing, $.list_literal), $._variable),
         seq($._expression, '=>', choice(alias($._list_destructing, $.list_literal), $._variable))
       ))),
+      optional(','),
```
Comment (maxbrunsfeld) [inline grammar.js:1021]: It seems like this `optional(',')` rule is adding ambiguity.

Do you need this, since the values inside the `commaSep1` are optional? I would think that a trailing comma would match the `commaSep1`.

File: grammar.js [hunk 2]
```diff
@@ -1010,19 +1014,21 @@ module.exports = grammar({
...
 
     _array_destructing: $ => seq(
       '[',
-      commaSep(choice(
+      commaSep1(optional(choice(
         choice(alias($._array_destructing, $.list_literal), $._variable),
         seq($._expression, '=>', choice(alias($._array_destructing, $.list_literal), $._variable))
-      )),
+      ))),
+      optional(','),
```
Comment (maxbrunsfeld) [inline grammar.js:1031]: I have the same question as above for this rule.

**Review by maxbrunsfeld — COMMENTED** [review]
This is great. I am wondering if it's possible to match all of this without the new conflicts.

File: grammar.js (hunk 1 above)
Comment (cfroystad) [reply grammar.js]: You're perfectly right! Thanks for stewarding this project and sharing your knowledge!

File: src/scanner.cc [hunk 3]
```diff
@@ -146,7 +146,8 @@ struct Scanner {
 
     for (;;) {
       if (position_in_word == heredoc.word.size()) {
-        if (lexer->lookahead == ';' || lexer->lookahead == '\n' || lexer->lookahead == '\r') {
+        // , and ) is needed to support heredoc in function arguments
+        if (lexer->lookahead == ';' || lexer->lookahead == ',' || lexer->lookahead == ')' || lexer->lookahead == '\n' || lexer->lookahead == '\r') {
```
Comment (cfroystad) [inline src/scanner.cc:150]: After thinking a bit more about this, I realize that while this works in the ideal case, we might run into problems if e.g. the user places a space before the comma or closing parenthesis.

Should I just add a special case checking all whitespace until the next char or is there some idiomatic tree-sitter way of handling this?

Comment (maxbrunsfeld) [reply src/scanner.cc:150]: Yeah, there's no built-in facility for this, so I think you have to manually skip over whitespace.

**Comment by maxbrunsfeld** [comment]
Thanks!
